Add support for PGO to rust, fix clang pgo

dcbaker opened the pull request:
Clang (and other LLVM compilers) handle PGO slightly differently than GCC based compilers do. The following is a short list of differences:
  1) LLVM compilers generate files that must be merged with llvm-profdata, GCC based compilers do not
  2) LLVM compilers do not generate profile data for static libraries, but GCC based compilers do
  3) LLVM compilers want an explicit path to put their profile data in, while GCC compilers put the data next to the .o files

Fixes #5251 
Fixes #13371

Based on #13377

dcbaker committed:
compilers: pass the private dir into profile_args

This is important for Rust, which needs to have a path to write
temporary files to, and we want to put that in the target's private
directory.

dcbaker committed:
compilers: Pass a private PGO dir to the pgo options

Rustc needs a directory to read and write files from, and clang can make
use of this as well (which may be useful for profiling mixed rustc and
clang targets).

dcbaker committed:
compilers/clang: Use explicit PGO directory

GCC and Clang are actually somewhat different than each other on PGO.
GCC, expects a single directory where data will be written, and will
read and write itself. Clang expects a directory where it will write
data as the binary runs, and then expects an external tool
(llvm-profdata) to merge the resulting instrumentation files before
clang uses the profile data.

dcbaker committed:
backends/ninja: Add a rule for merging raw LLVM PGO files

dcbaker committed:
compilers: add a base compiler method for getting the merged profile

This helps to reduce code duplication around the code base.

dcbaker committed:
backends/ninja: annotate the generate_single_compile function

dcbaker committed:
backends/ninja: rename the add_header_deps method

header dependency are usually order-only, but this actually provides a
full dependency.

dcbaker committed:
backends/ninja: cleanup dep handling a bit

dcbaker committed:
backends/ninja: fix dependencies added as full but should be order-only

Due to the confusing naming of the parameters, a number of arguments
were incorrectly added as full dependencies when they should be order
only.

dcbaker committed:
backends/ninja: Generate PGO rules correctly for clang

dcbaker committed:
compilers/rust: Add support for PGO

Rustc works like clang, so we just need to plug into that path.

dcbaker committed:
compilers: Use a compiler method to determine if PGO should happen

LLVM based compilers do not generate profile data for static libraries,
but GCC based compilers do.

tristan957 commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2947,14 +2947,24 @@ def generate_common_compile_args_per_src_type(self, target: build.BuildTarget) -
-        Compiles C/C++, ObjC/ObjC++, Fortran, and D sources
+        """Compiles C/C++, ObjC/ObjC++, Fortran, and D sources.
+
+        :param target: The target which the source belongs to
```
Be consistent about starting the description with a capital letter or not.

tristan957 commented on mesonbuild/compilers/compilers.py:
```diff
@@ -285,7 +285,16 @@ def are_asserts_disabled(options: KeyedOptionDictType) -> bool:
 
 
-def get_base_compile_args(options: 'KeyedOptionDictType', compiler: 'Compiler', env: 'Environment') -> T.List[str]:
+def get_base_compile_args(options: 'KeyedOptionDictType', compiler: 'Compiler', env: 'Environment', privatedir: str) -> T.List[str]:
```
I would name the parameter `private_dir` to match the target function name.

tristan957 commented on mesonbuild/compilers/mixins/gnu.py:
```diff
@@ -625,5 +625,5 @@ def use_linker_args(cls, linker: str, version: str) -> T.List[str]:
 
-    def get_profile_use_args(self) -> T.List[str]:
-        return super().get_profile_use_args() + ['-fprofile-correction']
+    def get_profile_use_args(self, priv_dir: str) -> T.List[str]:
```
Consistent parameter name

tristan957 commented on mesonbuild/compilers/compilers.py:
```diff
@@ -285,7 +285,9 @@ def are_asserts_disabled(options: KeyedOptionDictType) -> bool:
-def get_base_compile_args(options: 'KeyedOptionDictType', compiler: 'Compiler', env: 'Environment', privatedir: str) -> T.List[str]:
+def get_base_compile_args(options: 'KeyedOptionDictType', compiler: 'Compiler',
+                          env: 'Environment', privatedir: str,
+                          target: BuildTarget) -> T.List[str]:
```
Now that you are passing the target, it seems like you could remove privatedir.

tristan957 commented:
Is it possible to add any tests for PGO?

dcbaker committed:
tests: add a couple of pgo unittests

Since we don't seem to have any tests for this feature

dcbaker commented:
changes since the last version:
  - made the argument name consistently `pgo_dir`, since this directory is a subdirectory of the private dir
  - added a few tests for PGO, including mixing C and Rust
